crates/apollo-encoder/src/selection_set.rs

```diff
@@ -0,0 +1,205 @@
+use crate::{Field, FragmentSpread, InlineFragment};
+use std::fmt;
+
+/// The __selectionSet type represents a selection_set type in a fragment spread, an operation or a field
+///
+/// *SelectionSet*:
+///     Selection*
+///
+/// Detailed documentation can be found in [GraphQL spec](https://spec.graphql.org/October2021/#sec-Selection-Sets).
+///
+/// ### Example
+/// ```rust
+/// use apollo_encoder::{Field, FragmentSpread, Selection, SelectionSet, TypeCondition};
+/// use indoc::indoc;
+///
+/// let selections = vec![
+///     Selection::Field(Field::new(String::from("myField"))),
+///     Selection::FragmentSpread(FragmentSpread::new(String::from("myFragment"))),
+/// ];
+/// let mut selection_set = SelectionSet::new();
+/// selections
+///     .into_iter()
+///     .for_each(|s| selection_set.selection(s));
+///
+/// assert_eq!(
+///     selection_set.to_string(),
+///     indoc! {r#"
+///         {
+///           myField
+///           ...myFragment
+///         }
+///     "#}
+/// )
+/// ```
+#[derive(Debug, PartialEq, Clone, Default)]
+pub struct SelectionSet {
+    selections: Vec<Selection>,
+}
+
+impl SelectionSet {
+    /// Create an instance of SelectionSet
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Create an instance of SelectionSet given its selections
+    pub fn with_selections(selections: Vec<Selection>) -> Self {
+        Self { selections }
+    }
+
+    /// Add a selection in the SelectionSet
+    pub fn selection(&mut self, selection: Selection) {
+        self.selections.push(selection);
+    }
+
+    /// Should be used everywhere in this crate isntead of the Display implementation
+    /// Display implementation is only useful as a public api
+    pub(crate) fn format_with_indent(&self, mut indent_level: usize) -> String {
+        let mut text = String::new();
+        text.push_str("{\n");
+        indent_level += 1;
+        let indent = "  ".repeat(indent_level);
+        for sel in &self.selections {
+            text.push_str(&format!(
+                "{}{}\n",
+                indent,
+                sel.format_with_indent(indent_level)
+            ));
+        }
+        if indent_level <= 1 {
+            text.push_str("}\n");
+        } else {
+            text.push_str(&format!("{}}}", "  ".repeat(indent_level - 1)));
+        }
+
+        text
+    }
+}
+
+// This impl is only useful when we generate only a SelectionSet
+// If it's used from a root element, we call `format_with_indent`
```

Comment:
What do you mean by `root` here? The top-level `Document`?

Reply:
I should use `parent` instead

Reply:
yes `parent` makes more sense to me!